ua-snap/iem-webapp
# Handle cases where no data is received from API endpoint or invalid location is requested, and adapt report page accordingly

Closes #212.
Closes #189.
Closes #148.
Closes #136.
Closes #5.
Closes #245 
Xref: #176.
Xref: #142.

This PR does a couple of things:

- Handles HTTP error codes (missing data, invalid coordinates, server errors) and adapts report pages accordingly. Sections with missing data and their table of contents entries are hidden, the intro text adapts to what data is available, and the user is notified what data is missing and why (to the extent that it's possible from the HTTP error codes the web app is provided).
- If every API endpoint returns the same HTTP error code, which is often the case, all of the sections, intro text, and qualitative text are removed from the report. The only thing that remains is the title (name or coordinates of the location), mini-map, and the one HTTP error shared by all API endpoints (e.g., "No data at this location" or "Invalid coordinate").

To test this PR:

- Load a location like Fairbanks, which has temperature, precipitation, and permafrost data, but not wildfire data. Make sure there is no wildfire explanation text near the bottom of the page, and the table of contents entry for wildfire has been removed. Near the top of the report into text, there should be no implication that wildfire data exists. Instead, near the bottom of the intro text, it will say that wildfire data does not exist at this location any why.
- Select a location with no data for any API endpoint, like a point off the coast of Alaska. The easiest way to do this is to modify the lat/lon parameters manually in the URL. 
- Select a location that is outside of the valid bounding box (172.4201, 51.3492, -122.8098, 71.3694). Again, the easiest way to do this is to modify the lat/lon parameters manually in the URL. 

While developing this PR, I ran a modified version of the API that faked an HTTP error code for each of the endpoints. The report page adapts accordingly regardless of which API endpoint fails.
